test/pipeline_test_runner.cc

```diff
@@ -413,18 +413,40 @@ TEST_CASE("PerPhaseTest") { // NOLINT
     }
 
     {
+        // Immutable namer + resolver mode, which is used by things like serving LSP requests from stale GlobalStates
+
+        // Stashes any already-reported errors so we can check them later (we're going to toss all
+        // errors created at the end of this mode).
         handler.drainErrors(*gs);
 
+        // This mode is not designed to run on files that Sorbet did not already know about. In LSP
+        // mode this constraint is satisfied naturally by way of the implementation. Here, we have
+        // to satisfy it by forcing the file-level `<static-init>` methods into existence first.
+        // This would mutate the rest of the pipeline, so for the sake of not polluting things
+        // downstream from us, we do it on a copy of GlobalState.
+        auto staleGS = gs->deepCopy();
+        {
+            core::UnfreezeNameTable nameTableAccess(*staleGS);
+            core::UnfreezeSymbolTable symbolTableAccess(*staleGS);
+            for (auto &tree : trees) {
+                staleGS->staticInitForFile(core::Loc(tree.file, tree.tree.loc()));
+            }
+        }
+
+        const core::GlobalState &immutableGS = *staleGS;
+
         // Non-mutating namer, before entering symbols into GlobalState
         vector<ast::ParsedFile> treesCopy;
         for (auto &tree : trees) {
             treesCopy.emplace_back(ast::ParsedFile{tree.tree.deepCopy(), tree.file});
         }
 
-        move(namer::Namer::symbolizeTreesBestEffort(*gs, move(treesCopy), *workers).result());
+        treesCopy = move(namer::Namer::symbolizeTreesBestEffort(immutableGS, move(treesCopy), *workers).result());
         ENFORCE(!gs->hadCriticalError());
 
-        // If no ENFORCE fired, then non-mutating namer is working fine.
+        treesCopy = move(resolver::Resolver::runIncremental(immutableGS, move(treesCopy)).result());
+
+        // If no ENFORCE fired, then immutatable namer + resolver is working fine.
```

Comment:
I've just realized that we probably want to run DefLocSaver and typecheckOne on these resolved trees. Both of those operate on immutable GlobalState already, so it's just a question of do our namer and resolver passes do a good enough job at cleaning up the trees.

Reply:
I added this.

Reply:
Was the intent here to ensure that we're able to answer LSP queries when using a stale global state?

Reply:
yep!